_posts/2019-08-21-rust-bloat.markdown

```diff
@@ -0,0 +1,108 @@
+
+### Sharing
+
+I sometimes hear that it's ok to depend on commonly-used crates, because their cost is amortized among the various users that share them. I'm not convinced, for a variety of reasons. For one, it's common that you get different versions anyway (the [Zola] build currently has two versions each of unicase, parking_lot, parking_lot_core, crossbeam-deque, toml, derive_more, lock_api, scopeguard, and winapi). Second, if generics are used heavily (see below), there'll likely be code duplication anyway.
+
+That said, for stuff like Unicode data, it is quite important that there as few copies as possible in the binary. The best choice is crates engineered to be lean.
+
+### Proc macros
+
+A particularly contentious question is proc macros. The support crates for these (syn and quote) take maybe 10s to compile, and don't directly impact executable size. It's a major boost to the expressivity of the Rust language, and we'll likely use them in druid, though have been discussing making them optional.
+
+What I'd personally like to see is proc macros stabilize more and then be adopted into the language.
+
+## Use serialization sparingly
+
+Digging into xi-editor, the biggest single source of bloat is serde, and in general the fact that it serializes everything into JSON messages. This was something of an experiment, and in retrospect I would say one of the things I'm most unhappy about. It seems that efficient serialization is not a solved problem yet. [Note also that JSON serialization is [extremely slow](https://github.com/xi-editor/xi-mac/issues/102) in Swift]
+
+## Use polymorphism sparingly
+
+The *particular* reason serde is so bloated is that it monomorphizes everything. There are alternatives; [miniserde] in particular yields smaller binaries and compile times by using dynamic dispatch (trait objects) in place of monomorphization. But it has other limitations and so hasn't caught on yet.
+
+In general, overuse of polymorphism is a leading cause of bloat. For example, resvg [switched from lyon to kurbo](https://github.com/RazrFalcon/resvg/commit/708d0fff2bf47939587e0d562085a65f6dbf794f) for this reason. We don't the lyon / euclid ecosystem, also for this reason, which is something of a shame because now there's more fragmentation. When working on [kurbo], I did experiments indicating there was no real benefit to allowing floating point types other than `f64`, so just decided that would be the type for coordinates. I'm happy with this choice.
```

Comment:
third sentence is missing something, "We don't _____ the lyon..."

Reply:
adopt